url.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 Walter Scheper <walter.scheper@gmail.com>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package rapid
+
+import (
+	"fmt"
+	"net/url"
+	"path"
+	"reflect"
+	"strings"
+	"unicode"
+)
+
+type domainNameGen struct {
+	maxLength        int
+	maxElementLength int
+}
+
+func (g *domainNameGen) String() string {
+	return fmt.Sprintf("Domain(maxLength=%v, mmaxElementLength%v)", g.maxLength, g.maxElementLength)
+}
+
+func (g *domainNameGen) type_() reflect.Type {
+	return stringType
+}
+
+func (g *domainNameGen) value(t *T) value {
+	domain := SampledFrom(tlds).
+		Filter(func(s string) bool { return len(s)+2 <= g.maxLength }).
+		Map(func(s string) string {
+			var n string
+			for _, ch := range s {
+				n += string(SampledFrom([]rune{unicode.ToUpper(ch), unicode.ToLower(ch)}).Draw(t, "").(rune))
+			}
+
+			return n
+		}).Draw(t, "domain").(string)
+
+	var b strings.Builder
+	elements := newRepeat(1, 126, 1)
+	b.Grow(elements.avg())
+
+	var expr string
+	switch g.maxElementLength {
+	case 1:
+		expr = `[a-zA-Z]`
+	case 2:
+		expr = `[a-zA-Z][a-zA-Z0-9]?`
+	default:
+		expr = fmt.Sprintf(`[a-zA-Z]([a-zA-Z0-9\-]{0,%d}[a-zA-Z0-9])?`, g.maxElementLength-2)
+	}
+	for elements.more(t.s, g.String()) {
+		subDomain := StringMatching(expr).Draw(t, "subdomain").(string)
+		if len(domain)+len(subDomain) >= g.maxLength {
+			break
+		}
+		domain = subDomain + "." + domain
+	}
+
+	return domain
+}
+
+// Domain generates an RFC 1035 compliant domain name.
+func Domain() *Generator {
+	return DomainOf(255, 63)
+}
+
+// DomainOf generates an RFC 1035 compliant domain name,
+// with a maximum overall length of maxLength
+// and a maximum number of elements of maxElements.
+func DomainOf(maxLength, maxElementLength int) *Generator {
+	assertf(4 <= maxLength, "maximum length (%v) should not be less than 4, to generate a two character domain and a one character subdomain", maxLength)
+	assertf(maxLength <= 255, "maximum length (%v) should not be greater than 255 to comply with RFC 1035", maxLength)
+	assertf(1 <= maxElementLength, "maximum element length (%v) should not be less than 1 to comply with RFC 1035", maxElementLength)
+	assertf(maxElementLength <= 63, "maximum element length (%v) should not be greater than 63 to comply with RFC 1035", maxElementLength)
+
+	return newGenerator(&domainNameGen{
+		maxElementLength: maxElementLength,
+		maxLength:        maxLength,
+	})
+}
+
+type urlGenerator struct {
+	schemes []string
+}
+
+func (g *urlGenerator) String() string {
+	return fmt.Sprintf("URLGenerator(schemes=%v)", g.schemes)
+}
+
+func (g *urlGenerator) type_() reflect.Type {
+	return reflect.TypeOf(url.URL{})
+}
+
+func (g *urlGenerator) value(t *T) value {
+	scheme := SampledFrom(g.schemes).Draw(t, "scheme").(string)
+	domain := Domain().Draw(t, "domain").(string)
```

Comment:
Shouldn't we generate URLs with host component containing an IP (v4 or v6) address as well? That does not sound easy (especially given the IPv6 syntax variants), but otherwise we are omitting quite an important URL subset.

Reply:
Good question. If I add those options, should I do so via public IPv4 and IPv6 Generators?

Reply:
Yes, sounds good. Maybe a single generator of `net.IP` with an option like `v6 bool`?

Reply:
Adding the IPv4 and IPv6 generators is pretty straightforward, but I'm not sure how to incorporate them into the URL generator.

I can think of three ways to do this:
* One generator, `URL()`, that generates both DNS URLs and IP URLs
* Two generators, `URL()` that generates DNS URLs and `URLOf(schemes []string, domain *Generator)`.
* Three generators, `URL()`, `URLIPv4(), and `URLIPv6()`

I think you're looking for the first option, but wanted to float the other two and see what you think.

Reply:
I think going with just `URL()` for now is the safest option (because it can be extended in the future based on real feedback) -- in other cases we are guessing/speculating a bit about the use cases.

Reply:
Done.

Reply:
Great, thanks! I hope I'll be able to review/merge all this before the end of the week.